include/FilesystemManagement.h

```diff
@@ -0,0 +1,156 @@
+
+   However there are some practical restrictions, so if your project involves
+   any of the following you should contact us and we will try to work something
+   out:
+    - DRM or Copy Protection of any kind(except Copyrights)
+    - Software Patents You Do Not Wish to Freely License
+    - Any Kind of Linking to Non-GPL licensed Works
+    - Are Currently In Violation of Another Copyright Holder's GPL License
+    - If You want to change our code and not add a few hundred MB of stuff to
+        your distribution
+
+   These and other limitations could cause serious legal problems if you ignore
+   them, so it is best to simply contact us or the Free Software Foundation, if
+   you have any questions.
+
+   Joseph Toppi - toppij@gmail.com
+   John Blackwood - makoenergy02@gmail.com
+*/
+#ifndef Mezz_Filesystem_FilesystemManagement_h
+#define Mezz_Filesystem_FilesystemManagement_h
+
+#ifndef SWIG
+    #include "DataTypes.h"
+#endif
+
+namespace Mezzanine {
+namespace Filesystem {
+    /// @brief A Collection of possible results from common filesystem operations.
+    /// @remarks Any value larger than success should be considered an error/failure. Any function that returns
+    /// such a value did not complete and in most cases did not modify the filesystem.
+    enum class [[nodiscard]] ModifyResult
+    {
+        Success = 0,       ///< Everything worked.
+        // Existence Errors
+        AlreadyExists,     ///< The file or directory to be created already exists.
+        DoesNotExist,      ///< The file of directory specified wasn't found.
+        // Path Errors
+        InvalidPath,       ///< The path provided is invalid. Could be caused by a "." at the end.
+        LoopingPath,       ///< The path provided contains links that form a loop.
+        NameTooLong,       ///< The name of a file or directory specified is too long.
+        // Permission Errors
+        PermissionDenied,  ///< The filesystem permissions forbid the action to be taken.
+        ReadOnly,          ///< The filesystem only permits read operations.
+        // Directory Errors
+        NotADirectory,     ///< The path was expected to direct to a directory, but doesn't.
+        IsADirectory,      ///< The path was expected to direct to non-directory link, but does.
+        NotEmpty,          ///< The directory specified is not empty.
+        // Other Errors
+        IOError,           ///< There was a hardware I/O error while completing the operation.
+        NoSpace,           ///< There is insufficient space on the filesystem to complete the operation.
+        MaxLinksExceeded,  ///< The operation would exceeded the maximum links supported for the directory.
+        CurrentlyBusy,     ///< The file or directory specified is in use by the system and operation can't complete.
+        OperationCanceled, ///< The operation was canceled or aborted.
+        // Unknown
+        Unknown            ///< Error is unknown.
+    };
```

Comment:
How can we make an operator bool for this so it can be used directly inside and if statement consistently?

Reply:
We cannot.  But I have made a few equality and inequality operators.